subprojects/base-services/src/test/groovy/org/gradle/internal/hash/HashingTest.groovy

```diff
@@ -16,16 +16,81 @@
 
 package org.gradle.internal.hash
 
+
 import spock.lang.Specification
 
+import static java.lang.Thread.currentThread
+import static java.util.concurrent.CompletableFuture.supplyAsync
+import static java.util.concurrent.Executors.newFixedThreadPool
+
 class HashingTest extends Specification {
+    def 'cannot call hash multiple times'() {
+        given:
+        def hasher = Hashing.newHasher()
+        hasher.putInt(1)
+
+        when:
+        hasher.hash()
+        hasher.hash()
+
+        then:
+        thrown(IllegalStateException)
+    }
+
+    def 'nested hashers works'() {
```

Comment:
By nested I meant that we create A and B and call B.hash() before A.hash().
This was the reason I had to revert the `ThreadLocal`.
I'll make it clear in the test.

>  I think it's enough to test that newHasher() returns a new instance every time when called multiple times.

But that's not the source of the problem. The problem is that `new` and `hash` are not atomic.
If we were hashing e.g. `byte[]` only, like in `Hashing#hashBytes`, we could use `ThreadLocal`, since there would be no way for two hashers to be tangled.
That's why I'm testing the root cause, i.e. that if the `new` and `hash` methods mixed, we fail.
We could return the same instance and not fail in the above mentioned scenario.

Reply:
"Nested" for me means some outer thing fully encloses an inner thing, like in a nested inner class. The problem you are mentioning is related to this nesting being broken. I would probably call this case "overlapping", though that's not a perfect name either.

Anyway, the contract (reflected in the name of the method `newHasher()`) is that you get a new `Hasher` every time. Even testing that is a bit like testing the implementation directly, and thus kind of redundant. Doing it this way feels like we are testing for missing implementation (the problem arises if we tie the hasher to the thread, and we are not doing that), which seems even more redundant.